[WIP] Add kubevirt provider to IPI

nirarg opened the pull request:
**WORK IN PROGRESS**

nirarg committed:
Add kubevirt specific behavior for cluster assets generation

* Update tfvars with cluster's values
* Update metadata object, Saved for destroy operation

nirarg committed:
Add machine (for controlplane) and machineset (for compute) to KubeVirt provider

* Those configurations are used in the "manifests" stage
* Use providerSpec (declared in cluster_api_provider_kubevirt) inside the configurations

nirarg committed:
Add kubevirt rhcos image

* Use the same image used by openstack and ovirt

bardielle commented on pkg/types/kubevirt/defaults/platform.go:
```diff
@@ -0,0 +1,11 @@
+package defaults
```
remove that file from that commit

bardielle commented on pkg/types/kubevirt/machinepool.go:
```diff
@@ -0,0 +1,44 @@
+
+	SourcePvcName string `json:"SourcePvcName,omitempty"`
+
+	SecretName string `json:"SecretName.omitempty"`
```
Call it the same as in KubevirtMachinerProviderSpec - UnderKubeconfigSecretName

bardielle commented on pkg/types/kubevirt/machinepool.go:
```diff
@@ -0,0 +1,44 @@
+	// Memory is the size of a VM's memory.
+	// Format: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/api/resource/quantity.go
+	// +optional
+	Memory string `json:"memory,omitempty"`
```
RequestedMemory

bardielle commented on pkg/types/kubevirt/machinepool.go:
```diff
@@ -0,0 +1,44 @@
+type MachinePool struct {
+	// CPU is the mount of cpus used
+	// +optional
+	CPU uint32 `json:"cpu,omitempty"`
```
RequestedCPU

bardielle commented on pkg/types/kubevirt/machinepool.go:
```diff
@@ -0,0 +1,44 @@
+	// StorageSize is the size of VM's boot volume
+	// Format: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/api/resource/quantity.go
+	// +optional
+	StorageSize string `json:"storageSize,omitempty"`
```
RequestedStorage

bardielle commented on pkg/types/kubevirt/machinepool.go:
```diff
@@ -0,0 +1,44 @@
+
+	SecretName string `json:"SecretName.omitempty"`
+
+	// TODO kubevirt add more params
```
Add more params and remove that comment

bardielle commented on pkg/types/kubevirt/metadata.go:
```diff
@@ -0,0 +1,7 @@
+
+// Metadata contains kubevirt metadata (e.g. for uninstalling the cluster).
+type Metadata struct {
+	KubeConfig    string `json:"kubeconfig"`
```
UnderKubeconfig

bardielle commented on pkg/types/kubevirt/platform.go:
```diff
@@ -0,0 +1,25 @@
+// machinesets use.
+type Platform struct {
+	// The path to the kubeconfig file of the underkube cluster
+	KubeConfig string `json:"kubevirt_kubeconfig"`
```
UnderKubeconfig

bardielle commented on pkg/types/kubevirt/platform.go:
```diff
@@ -0,0 +1,25 @@
+package kubevirt
+
+// Platform stores all the global configuration that all
```
Change the comment to be: holds all the platform information

bardielle commented on pkg/types/kubevirt/platform.go:
```diff
@@ -0,0 +1,25 @@
+	// own platform configuration.
+	// Default will set the image field to the latest RHCOS image.
+	// +optional
+	DefaultMachinePlatform *MachinePool `json:"defaultMachinePlatform,omitempty"`
```
Update the comment

bardielle commented on pkg/types/kubevirt/validation/machinepool.go:
```diff
@@ -0,0 +1,67 @@
+}
+
+// ValidVMType returns true if the vmType is supported.
+func ValidVMType(vmType ovirt.VMType) bool {
```
Remove that function

bardielle commented on pkg/types/kubevirt/validation/machinepool.go:
```diff
@@ -0,0 +1,67 @@
+}
+
+// supportedVMTypes returns a slice of all supported VMTypes.
+func supportedVMTypes() []ovirt.VMType {
```
Remove that function too

bardielle commented on pkg/types/kubevirt/validation/machinepool_test.go:
```diff
@@ -0,0 +1,88 @@
+	"github.com/openshift/installer/pkg/types/ovirt"
+)
+
+func TestValidateMachinePool(t *testing.T) {
```
Add a todo comment to add more unit-tests

bardielle commented on pkg/types/kubevirt/validation/platform.go:
```diff
@@ -0,0 +1,34 @@
+func ValidatePlatform(p *ovirt.Platform, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if err := validate.UUID(p.ClusterID); err != nil {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("ovirt_cluster_id"), p.ClusterID, err.Error()))
```
change it to be kubevirt_cluster_id

bardielle commented on pkg/types/kubevirt/validation/platform.go:
```diff
@@ -0,0 +1,34 @@
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("ovirt_cluster_id"), p.ClusterID, err.Error()))
+	}
+	if err := validate.UUID(p.StorageDomainID); err != nil {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("ovirt_storage_domain_id"), p.StorageDomainID, err.Error()))
```
change it to be kubevirt_cluster_id

bardielle commented on pkg/types/kubevirt/validation/platform.go:
```diff
@@ -0,0 +1,34 @@
+)
+
+// ValidatePlatform checks that the specified platform is valid.
+func ValidatePlatform(p *ovirt.Platform, fldPath *field.Path) field.ErrorList {
```
update that function to be relevant to kubevirt platform

bardielle commented on pkg/types/kubevirt/validation/platform_test.go:
```diff
@@ -0,0 +1,109 @@
+)
+
+func validPlatform() *ovirt.Platform {
+	return &ovirt.Platform{
```
update it with validations

bardielle commented on pkg/types/kubevirt/validation/platform_test.go:
```diff
@@ -0,0 +1,109 @@
+	}
+}
+
+func TestValidatePlatform(t *testing.T) {
```
Add a todo comment to add more tests

bardielle commented on pkg/types/ovirt/doc.go:
```diff
@@ -3,4 +3,4 @@
 
 // Name is the name for the ovirt platform.
-const Name string = "ovirt"
+const 	Name string = "ovirt"
```
Why did you change that file?

bardielle commented on pkg/tfvars/kubevirt/kubevirt.go:
```diff
@@ -0,0 +1,61 @@
+
+type config struct {
+	Namespace        string `json:"kubevirt_namespace"`
+	Kubeconfig       string `json:"kubevirt_kubeconfig"`
```
UnderKubeconfig

bardielle commented on pkg/tfvars/kubevirt/kubevirt.go:
```diff
@@ -0,0 +1,61 @@
+	RequestedMemory  string `json:"kubevirt_master_memory"`
+	RequestedCPU     uint32 `json:"kubevirt_master_cpu"`
+	RequestedStorage string `json:"kubevirt_master_storage"`
+	MachineType      string `json:"kubevirt_master_machine_type"`
```
Remove that type

bardielle commented on pkg/tfvars/kubevirt/kubevirt.go:
```diff
@@ -0,0 +1,61 @@
+	Kubeconfig       string `json:"kubevirt_kubeconfig"`
+	ImageURL         string `json:"kubevirt_image_url"`
+	SourcePvcName    string `json:"kubevirt_source_pvc_name"`
+	SecretName       string `json:"kubevirt_secret_name"`
```
SecretName of what? I think you need to remove it

bardielle commented on pkg/tfvars/kubevirt/kubevirt.go:
```diff
@@ -0,0 +1,61 @@
+// TFVarsSources contains the parameters to be converted into Terraform variables
+type TFVarsSources struct {
+	MasterSpecs []*v1.KubevirtMachineProviderSpec
+	Kubeconfig  string
```
UnderKubeconfig

bardielle commented on pkg/tfvars/kubevirt/kubevirt.go:
```diff
@@ -0,0 +1,61 @@
+type TFVarsSources struct {
+	MasterSpecs []*v1.KubevirtMachineProviderSpec
+	Kubeconfig  string
+	ImageURL    string
```
ImageUrl of what? rchoseimageUrl

bardielle commented on pkg/terraform/exec/plugins/kubernetes-alpha.go:
```diff
@@ -0,0 +1,58 @@
+	"google.golang.org/grpc"
+)
+
+// func init() {
```
remove that comment

bardielle commented on pkg/terraform/exec/plugins/kubernetes-alpha.go:
```diff
@@ -0,0 +1,58 @@
+		ProtocolVersion: 5,
+		// The magic cookie values should NEVER be changed.
+		MagicCookieKey:   "TF_PLUGIN_MAGIC_COOKIE",
+		MagicCookieValue: "d602bf8f470bc67ca7faa0386276bbdd4330efaf76d1a219cb4d6991ca9872b2",
```
Use consts

bardielle replied:
still relevant

bardielle commented on pkg/terraform/exec/plugins/kubernetes-alpha.go:
```diff
@@ -0,0 +1,58 @@
+// 	provider.Serve()
+// }
+
+func init() {
```
Add a comment of description

bardielle commented on pkg/terraform/exec/plugins/kubernetes.go:
```diff
@@ -0,0 +1,14 @@
+		plugin.Serve(&plugin.ServeOpts{
+			ProviderFunc: kubernetes.Provider})
+	}
+	KnownPlugins["terraform-provider-kubernetes"] = exec
```
use consts

bardielle replied:
still relevant

bardielle commented on pkg/asset/installconfig/basedomain.go:
```diff
@@ -61,6 +62,9 @@ func (a *baseDomain) Generate(parents asset.Parents) error {
 			return err
 		}
+	case kubevirt.Name:
+		a.BaseDomain = "svc.cluster.local"
```
Use consts

bardielle commented on pkg/asset/installconfig/basedomain.go:
```diff
@@ -61,6 +62,9 @@ func (a *baseDomain) Generate(parents asset.Parents) error {
 		if !(gcpconfig.IsForbidden(err) || gcpconfig.IsThrottled(err)) {
 			return err
 		}
+	case kubevirt.Name:
```
Add a todo -use a variable from user

bardielle commented on pkg/asset/installconfig/installconfig.go:
```diff
@@ -187,5 +199,8 @@ func (a *InstallConfig) platformValidation() error {
 	if a.Config.Platform.OpenStack != nil {
 		return icopenstack.Validate(a.Config)
 	}
+	if a.Config.Platform.Kubevirt != nil {
```
change it to switch-case?

bardielle commented on pkg/asset/installconfig/kubevirt/config.go:
```diff
@@ -0,0 +1,11 @@
+
+import "io/ioutil"
+
+func LoadUnderKubeConfig(filePath string) ([]byte, error) {
```
LoadUnderKubeconfig

nirarg committed:
Add kubevirt platform manifests configurations

* Use provider Infra cluster kubeconfig to generate the following secrets:
  - cloud-creds-secret.yaml template
  - role-cloud-creds-secret-reader.yaml template
* Add kubevirt platform parameters to cloudproviderconfig, include: namespace
* Skip DNS config and its CRD generation for kubevirt platform
* Modify bootstrap bootkube.sh

data/data/bootstrap

nirarg committed:
Kubevirt platform infrastructure and node network specific configurations (multus network)

* Set APIVIP for ignition access
* Set mcscertkey for APIVIP

bardielle commented on data/data/kubevirt/main.tf:
```diff
@@ -0,0 +1,68 @@
+provider "kubernetes" {
+  config_path = "/home/nargaman/.kube/config"
```
change the path to be relative for the running directory

nirarg replied:
Done

bardielle commented on data/data/kubevirt/main.tf:
```diff
@@ -0,0 +1,68 @@
+
+provider "kubernetes-alpha" {
+  server_side_planning = false
+  config_path = "/home/nargaman/.kube/config"
```
change the path to be relative for the running directory

nirarg replied:
Done

bardielle commented on pkg/asset/installconfig/kubevirt/kubevirt.go:
```diff
@@ -0,0 +1,160 @@
+	var selectedKubeConfig string
+
+	defaultValue := ""
+	home := os.Getenv("HOME")
```
In my opiniion you need to take the kubeconfig from a relative place in the running directory

nirarg replied:
Done

bardielle commented on data/data/manifests/openshift/role-cloud-creds-secret-reader.yaml.template:
```diff
@@ -14,6 +14,8 @@ metadata:
 {{- else if .CloudCreds.Ovirt}}
   name: ovirt-creds-secret-reader
+{{- else if .CloudCreds.Kubevirt}}
+  name: kubevirt-creds-secret-reader
```
what is it for?

bardielle commented on pkg/asset/ignition/machine/node.go:
```diff
@@ -36,6 +37,8 @@ func pointerIgnitionConfig(installConfig *types.InstallConfig, rootCA []byte, ro
 			ignitionHost = net.JoinHostPort(installConfig.VSphere.APIVIP, "22623")
 		}
+	case kubevirttypes.Name:
+		ignitionHost = net.JoinHostPort(installConfig.Kubevirt.APIVIP, "22623")
```
I think that this section could be refactored: 
`switch installConfig.Platform.Name() {
case baremetaltypes.Name:
	apiVipParam = installConfig.BareMetal.APIVIP
....
}
ignitionHost = net.JoinHostPort( apiVipParam , "22623")
`


But I'm not sure

nirarg committed:
Add kubevirt platform gather info implementation (GatherKubevirt)

nirarg committed:
Add the apiVIP to resolv.conf inside the bootstrab VM

This is done in order to allow the bootstrap to notify the cluster when it done

nirarg committed:
Adding InfraCluster client, used by validations and destroy

nirarg committed:
Add destroy implementation for kubevirt platform

nirarg committed:
Add kubevirt platform defaults

nirarg committed:
Add kubevirt platform validations

nirarg committed:
Go modules

nirarg committed:
Add terraform plugins for "kubernetes" and "kubevirt" resources

nirarg committed:
Add reviewers and approvers lists for kubevirt alias

nirarg committed:
Add terraform data files for kubevirt provide

nirarg committed:
Add kubevirt provider tfvars

nirarg committed:
Add Kubevirt platform structs

* types/kubevirt/machinepool - the struct which responsible to hold all infra cluster values which are may be different between controlPlane (masters) and compute (workers)
* types/kubevirt/platform - the struct which responsible to hold all infra cluster values which are common between controlPlane (masters) and compute (workers)
* types/kubevirt/metadata - the struct which resposible to save the critical values for the destroy operation

Auto generated file changes: data/data/install.openshift.io_installconfigs.yam

nirarg committed:
Add install-config logic for kubevirt platform

* Add specific parameters to request from user
* Add an option to request "Mobile Network CIDR" from user, enabled only in kubevirt platform
